src/recording/macromanager.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <QObject>
+
+#include "engine/enginemaster.h"
+#include "mixer/playermanager.h"
+#include "preferences/usersettings.h"
+#include "recording/recordingmanagerbase.h"
+
+#define MAX_MACRO_SIZE 1000
+#define NO_DECK nullptr
+
+struct MacroAction {
+    MacroAction(){};
+    MacroAction(double position, double target)
+            : position(position), target(target){};
+    double position;
+    double target;
+};
+
+struct Macro {
+    int length = 0;
+    MacroAction actions[MAX_MACRO_SIZE];
+    QString deck = NO_DECK;
+
+    Macro() {
+        qDebug() << "Constructing Macro";
+        std::fill_n(actions, MAX_MACRO_SIZE, MacroAction());
+    }
+
+    void appendHotcueJump(double origin, double target) {
+        qDebug() << "Macro: Appending jump";
+        actions[length].position = origin;
```

Comment:
- The length must be incremented atomically before writing into the reserved array slot. Otherwise the same slot could be used twice.
- Length check is missing
- The qDebug() will cause drop outs when called from the audio thread

Reply:
I'm not sure we need any atomics, at least not as members of this class. A Macro should never be used accessed by two threads simultaneously. It should be created by the main thread and then used by the audio thread, but only the audio thread should modify it. This is what we do for EffectProcessors and they don't have any atomic members.

Reply:
Effects are per-deck, right? On the other hand, the macro recorder is shared across all decks. AFAIK we have a separate thread per deck, to it could be accessed by multiple threads. As soon as the first jump has been recorded, the recorder should be disabled for the other decks, so we need atomic at least for the "assign macro recorder to a single deck" logic. And when in doubt, we should also use atomic elsewhere. Better safe than sorry.

Reply:
Yes, the macro recorder is shared across decks but only one deck records at a time. We have separate threads for reading audio from files for each deck, but the audio processing is all done in the audio thread by EngineBuffer. Determining which deck is recording should be done in the audio thread, so I don't think this is a risk of a race condition for that.

Reply:
OK. It would be good if we could make sure that stuff is always run in a certain thread. Can we assert which thread the code is running in?

Reply:
In this case we need to ensure that access to the array + length is mutual exclusive. Always!! The MacroManager **must not** access those members while the audio thread might ever do it. PS: This is a very brittle design that I would not recommend to any client.

I realized that otherwise we would try to implement a lock-free queue here. In that case we should use a proper implementation instead.

Reply:
Encapsulating the fixed size buffer into a separate class and passing a pointer atomically almost works. But you cannot atomically read the pointer and add a new element at the same time. The pointer might get reset while just adding an element.

Reply:
> The MacroManager must not access those members while the audio thread might ever do it.

Could we accomplish this by somehow signaling from the main thread to the audio thread to stop recording? Once that signal has been sent, would it be safe for the other thread to read the recorded macros?

Reply:
Maybe we can use [Farbot's RealtimeMutatable class](https://github.com/hogliux/farbot/blob/master/RealtimeMutatable.hpp)?

Reply:
Hmm, I'm not sure Farbot's RealtimeMutatable class would be appropriate. I think that is designed for cases where the realtime thread generates data on its own, not modifying a data structure shared with the nonrealtime thread.

Reply:
Perhaps a FIFO is the safest option after all.

Reply:
> Encapsulating the fixed size buffer into a separate class and passing a pointer atomically almost works. But you cannot atomically read the pointer and add a new element at the same time. The pointer might get reset while just adding an element.

I propose this design:

0. Add a std::atomic<int> member to EngineMaster to track the macro recording state which is set to a 3 state enabled/disabling/disabled enum.
1. When the macro recording ControlObject is enabled, create the Macro object on the main thread and pass it to EngineMaster. EngineMaster would set its atomic macro recording state to enabled. This should never override the pointer the audio thread is using because macro recording could only be enabled if it was disabled before. I think that would require using ControlObject::connectValueChangeRequest.
2. When the macro recording ControlObject is disabled, set EngineMaster's atomic macro recording state to disabling. If any hotcues are activated, record them in this last cycle of the audio thread. At the end of EngineMaster::process, set the atomic macro recording state to fully disabled.
3. The nonrealtime thread polls EngineMaster's macro recording state and does not access the Macro object until the state is fully disabled. This polling should only need to last for one (or two?) cycles of the audio thread. When macro recording is fully disabled, the Macro could be saved to the database.

Reply:
@xerus2000 @uklotzde what do you think?

Reply:
Sounds good, I'll start implementing and then get back :)

Reply:
If we wouldn't have to pass around the pointer for the Macro, we could finally decouple everything again... I don't like it when objects peer into members of objects they have no proper business with.